pubsub/pubsub.go

```diff
@@ -623,30 +640,33 @@ func (s *Subscription) Receive(ctx context.Context) (_ *Message, err error) {
 			})
 			return m2, nil
 		}
-		// A call to ReceiveBatch must be in flight. Wait for it.
-		waitc := s.waitc
-		s.mu.Unlock()
+		s.mu.Unlock() // unlock to allow message or error processing from background goroutine
 		select {
-		case <-waitc:
-			s.mu.Lock()
-			// Continue to top of loop.
 		case <-ctx.Done():
 			s.mu.Lock()
 			return nil, ctx.Err()
+		default:
+			// Continue to top of loop.
```

Comment:
Doesn't this spin-loop while it's waiting? I think you've removed any use of `s.waitc`.

Maybe you can put back the `waitc` case, and up above every time we get some messages, write something to `waitc`? Only close it / set to null when all batches are complete.